### ISSUE TITLE
Prohibit deleting a branch that contains stashed changes

### ISSUE THREAD

**Issue by 1st** [desktop/desktop#13714]
### Describe the feature or problem you’d like to solve

Very often I switch between branches, and sometimes I do not publish my branch until I did something useful. And when I switch to another branch to do my work, I stash changes in he old branch.

But it happened to me multiple times that I decided to delete a branch (that is not published on the remote) and it contains stashed changes. And ops - I need these stashed changes, but now they are gone, and I need to remember the config adjustments or code changes that I wanted to keep, but that was gone when I delete the branch with all the stashed changes.

### Proposed solution

I think that it will be beneficial to keep the stashed changes in either way:
- prevent deleting of a branch that contains any stashed changes. I will need manually to restore stashed changes and revert them. This way it minimises risks of losing any useful changes that I need to keep
- move the stashed changes to a new branch (to master by default). But in this case it's very important to do not override already stashed changes in the master branch (because we will lose other stashed changes that was before in the master branch). In this case we just need to ad stashed changes to a stack of the stashed changes, or combine changes together

### Additional context

This way it will be much easier to be confident that changes that are useful but you don't want to commit them - will stay with you as long as you want. Even if it's local configuration files that should be useful only personally for you.

Thanks.

**Comment by rlaphoenix** [comment]
I highly +1 this, I almost lost crucial data, and lost days of effort stemming from poor user experience surrounding stashing.

1. **Pull;** `origin:master` has been updated, Great! Let's pull the data.
2. **Conflict;** Uh-oh! Changes would conflict with current working directory changes. Deal with that first by stashing or such.
3. **Existing Stash;** I already have some changes stashed on master and there's no way to append to a stash.
4. **Stash Elsewhere;** Check out another branch e.g., `feat-patch` and `Bring changes to branch`. I stashed the changes here.
5. **Pull;** Go back to the `master` branch and Pull from origin, then merged feat-patch into master.
6. **Delete Branch;** Once merged, I deleted the `feat-patch` branch with no warning that there was a stash within it.

Of course, there is user error here to have deleted said branch, but one can very easily forget the stash was within it, especially in the mindset of having just merged its contents with another branch. It's also not that out there to expect a warning when similar warnings are done throughout the GitHub Desktop experience.

**Comment by tidy-dev** [comment]
Just for anyone running into this, your stash is not actually deleted. If you go to a terminal and run `git stash list`. You should be able to find it there, and pop it onto any new branch you would like.

**Comment by 1st** [comment]
> Just for anyone running into this, your stash is not actually deleted. If you go to a terminal and run `git stash list`. You should be able to find it there, and pop it onto any new branch you would like.

Thanks for the hint. But it's more a workaround than a proper workflow. Will be good to prevent a shot to your leg, instead of reanimating your leg after the shot. Not all people will end up to your message in this thread to restore the missing stashed changes (I lost mine couple of times, when deleting a branch that is merged and is deleted on remote). Potentially this small "bug" can lose a lot of time of many developers who will be at least once in such a situation. Better to save lifetime and invest it to something useful, instead of searching for a way how to restore a lost changed files. Agree?

**Comment by tidy-dev** [comment]
Absolutely! Just until this can be looked at - I wanted to provide the workaround for anyone looking.

**Comment by rlaphoenix** [comment]
Hi, I just wanted to mention that renaming a branch also removes the stash. Not ideal at all.
